### ISSUE TITLE
Third party libraries such as psycopg2 marked as W7936(missing-manifest-dependency)

### ISSUE THREAD

**Issue by guewen** [OCA/pylint-odoo#247]
I tracked down an issue with a check returning erroneous result:

```
odoo/local-src/camptocamp_tools/utils.py:10: [W7936(missing-manifest-dependency), ] Missing external dependency "psycopg2" from manifest. More info: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#external-dependencies
```

It happens here:

https://github.com/OCA/pylint-odoo/blob/67fa4893ce04b93e19612a577bc4563f0bfa62dc/pylint_odoo/checkers/modules_odoo.py#L356-L361

We add the list of modules to whitelist, psycopg2 is in the list, so we shouldn't have this warning. However, if you use `isort` to actually sort the imports, you may have a configuration file `.isort.cfg`. In this configuration, you will have the list of modules to consider as third party, so isort places them at the right place. Example:

```
[settings]
known_third_party = dateutil,openupgradelib,pkg_resources,psycopg2,requests,setuptools,urllib2,yaml
```

`isort.place_module()` builds a list of known patterns, and check first if the lib is a third party, returning this result first (if `psycopg2` is in the third party list and the std list, it will be considered as third party).

Refs (not in the most recent version, but easy to read): 
* https://github.com/timothycrosley/isort/blob/88a23e8aaacddf727ce01040dab31b9f7f55c497/isort/isort.py#L148-L155
* https://github.com/timothycrosley/isort/blob/88a23e8aaacddf727ce01040dab31b9f7f55c497/isort/isort.py#L257-L262

**Comment by moylop260** [comment]
Hi @guewen 
I have checked and `pyscopg2` is a import name whitelist:

 - https://github.com/OCA/pylint-odoo/blob/67fa4893ce04b93e19612a577bc4563f0bfa62dc/pylint_odoo/checkers/modules_odoo.py#L176

It is used from config parameter `import_name_whitelist`:
 - https://github.com/OCA/pylint-odoo/blob/67fa4893ce04b93e19612a577bc4563f0bfa62dc/pylint_odoo/checkers/modules_odoo.py#L220

I just tested the parameter for odoo addons where is used the same library.
`rgrep "import psycopg2" ./addons/*/models`
```bash
./addons/base_import/models/base_import.py:14:import psycopg2
./addons/delivery/models/delivery_carrier.py:5:import psycopg2
./addons/mail/models/mail_mail.py:7:import psycopg2
./addons/point_of_sale/models/pos_order.py:7:import psycopg2`
```

`pylint --load-plugins=pylint_odoo -d all -e missing-manifest-dependency ./addons/point_of_sale`
The result was empty.

Could you check your pylint configuration file if the parameter `import_name_whitelist` is re-defined?
Could you share us pylint, pylint-odoo and python version?

**Comment by guewen** [comment]
@moylop260 Thanks for checking. Indeed, `psycopg2` is in the whitelist. The problem is with the usage `pylint-odoo` makes of `isort`. It uses it to get a classification of each lib between "third party" and "std lib". The whitelist forces it to be a stdlib usually.

However, if you uses `isort` in your project, you'll have a `.isort.cfg` file in your project's folder, and in this file, you might have for instance:

```
[settings]
known_third_party = dateutil,openupgradelib,pkg_resources,psycopg2,requests,setuptools,urllib2,yaml
```

If you don't, `isort` will not put psycopg2 at the right place.

Having this file modifies the behavior of `pylint-odoo` because now, `isort` will classify `psycopg2` as third party instead of forcing it to `std lib`.

**Comment by moylop260** [comment]
@guewen 
I got it
Thank for explanation.

I just reproduced using:
```sh
echo """[settings]
known_third_party = psycopg2
""" > ~/.isort.cfg

pylint --load-plugins=pylint_odoo -d all -e missing-manifest-dependency ./addons/point_of_sale
```

Output:
`************* Module point_of_sale.models.pos_order
./addons/point_of_sale/models/pos_order.py:7:0: W7936: Missing external dependency "psycopg2" from manifest. More info: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#external-dependencies (missing-manifest-dependency)`

FYI I didn't find a easy fix since that the isort 4.3.4 is assigning the patterns to identify each library from the constructor `__init__` 😞:
 - https://github.com/timothycrosley/isort/blob/88a23e8aaacddf727ce01040dab31b9f7f55c497/isort/isort.py#L148-L155

And here is not easy to re-build them.

Newer versions 4.3.21 is using a separated class 😃 
 - https://github.com/timothycrosley/isort/blob/cd9c2f6e70196bd429d0179635e0fac2d034ac5e/isort/finders.py#L74-L86
Here it is good to re-build them.

But if we adds a patch just for newer versions locally will be raise an error for older ones.
Then it is more complicated than I expected to fix it.

**Comment by guewen** [comment]
Coming back on this, since it's becoming an issue with the new pre-commit configuration on 13.0 branches.
Do we have any reason to keep compatibility with `isort` versions older than 4.3.21?

**Comment by sbidoul** [comment]
Interesting, I did not know pylint-odoo was using isort..

**Comment by moylop260** [comment]
pylint is using isort and eslint.
But, Now black is using has integrated these features

Maybe we should delete these checks from pylint-odoo since that:
 - eslint is duplicated with black
 - isort is used to detect external packages without try-except and without a correct definition from manifest, but after https://github.com/odoo/odoo/pull/25549 it is not valid now.

So, for old odoo version we can use a old pylint-odoo version.
For odoo v13.0 or black implementation we can use newest pylint-odoo version.
